xla/service/gpu/gemm_algorithm_picker_test.cc

```diff
@@ -52,31 +53,101 @@ class GemmAlgorithmPickerTest : public HloTestBase,
-  if (comp.IsAtLeast(se::CudaComputeCapability::AMPERE)) {
-    GTEST_SKIP() << "Skipping this test for Ampere+ as it is supported and "
+  void SetUp() override {
+    std::visit(VariantVisitor{
+      [](const se::CudaComputeCapability& cc) {
+        if(cc.IsAtLeastAmpere()) {
+          GTEST_SKIP() << "Skipping this test for Ampere+ as it is supported and "
                     "recommended with "
                     "the Nvidia Volta+ GPUs.";
+        }
+      },
+      [this](const se::RocmComputeCapability& cc) {
+        if(GetDebugOptionsForTest().xla_gpu_enable_cublaslt() &&
+          !cc.has_hipblaslt()) {
+          GTEST_SKIP() << "No gpublas-lt support on this architecture!";
+        }
+      }},
+      gpu_comp());
+  }
+};
+
+TEST_P(GemmAlgorithmPickerTest, SkipAlgorithmsWithAccuracyCheck) {
+
+  constexpr absl::string_view kHlo = R"(
+HloModule module
+
+ENTRY main {
+  %arg0 = f32[100,100]{1,0} parameter(0)
+  %arg1 = f32[100,100]{1,0} parameter(1)
+  ROOT %dot = f32[100,100]{1,0} dot(arg0, arg1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
+})";
+
+  auto module_cfg = GetModuleConfigForTest();
+  auto debug_opts = module_cfg.debug_options();
+  size_t num_left1 = 0, num_left2 = 0;
+
+TF_ASSERT_OK_AND_ASSIGN(auto module,
+                          ParseAndReturnVerifiedModule(kHlo, module_cfg));
+
+  {
+    // Run first with default settings (autotune level = 4), keep the number of
+    // algorithms left after autotuning
+    TF_ASSERT_OK_AND_ASSIGN(
+      bool changed,
+      RunHloPass(
+          GemmRewriter(gpu_comp(), /*toolkit_version=*/12040),
+          module.get()));
+
+    AutotuneConfig cfg{DeviceConfig{stream_exec(), nullptr}, debug_opts};
+    GemmAlgorithmPicker gpicker(cfg);
+    TF_ASSERT_OK_AND_ASSIGN(changed, RunHloPass(gpicker, module.get()));
+    ASSERT_TRUE(changed);
+    num_left1 = gpicker.num_algorithms_left();
+    if(num_left1 < 2) {
+      GTEST_SKIP() << "Too few algorithms left after the first step";
+    }
+  }
+
+  // Clear cache before the second run!
+  AutotunerUtil::ClearAutotuneResults();
+  {
+    // Run once again but now with autotune level 5 and embarassingly tight 
+    // rtol which shall disqualify most of the algorithms.
+
+    // Note that, we have "two sources of truth" for GemmAlgorithmPicker: i.e.,
+    // debug_options are used to initialize both 'HloModuleConfig' and also 
+    // 'AutotuneConfig'.
+    debug_opts.set_xla_gpu_autotune_gemm_rtol(1e-12);
+    debug_opts.set_xla_gpu_autotune_level(5);
+    module->mutable_config().set_debug_options(debug_opts);
+    TF_ASSERT_OK_AND_ASSIGN(
+      bool changed,
+      RunHloPass(
+          GemmRewriter(gpu_comp(), /*toolkit_version=*/12040),
+          module.get()));
+
+    AutotuneConfig cfg{DeviceConfig{stream_exec(), nullptr}, debug_opts};
+    GemmAlgorithmPicker gpicker(cfg);
+    TF_ASSERT_OK_AND_ASSIGN(changed, RunHloPass(gpicker, module.get()));
+    ASSERT_TRUE(changed);
```

Comment:
I'm seeing this assertion flakily fail on V100s. I think the issue is the algorithm doesn't necessarily change the second time you call `RunHloPass`, since the selected algorithm from the first run of GemmAlgorithmPicker may have chosen an algorithm that doesn't have accuracy issues.

Probably the best way to fix this is to copy the module before passing it to GemmAlgorithmPicker, so you can pass a fresh copy to both GemmAlgorithmPicker instances.

Reply:
yes, I have tried running this test on V100 machine and removed **ASSERT_TRUE(changed);**. Not it seems to work fine: anyway, we do not care if GemmAlgorithmPicker has selected another algorithm: all that matters is that the number of algorithms used for autotuning during the second run (when **xla_gpu_autotune_level** is set to 5) is smaller that that of during the 1st run